File: examples/stream-publish-bot/ts/src/publish.ts
/**
 * OctoChat stream PUBLISH — a standalone Node script that publishes ONE message
 * into a PUBLIC stream room as a bot, then exits. No `/events`, no webhook, no
 * waiting: just the "post" half of an integration.
 *
 *     publish  ──append──▶  streampub room (access:'public')
 *               (audience-cap bot token)
 *
 * A stream room is an append-only log, so posting is a single signed `POST /push`
 * — no pull / merge / hash, no sync protocol. The bot is authorized by the
 * owner-minted "Connect a bot" token, a Starfish `createPublicLink` AUDIENCE cap
 * (no embedded secret): the bot generates its OWN keypair and signs the request
 * with it, naming that key via `X-Starfish-Pub` (`redeemPublicLink`). A leaked
 * token is therefore useless to anyone who can't also sign, and writes stay
 * attributable per bot.
 *
 * We reproduce `StarfishClient.append`'s exact wire format
 * (`body = JSON.stringify({ data })`, `Content-Type: application/json`) so the
 * server accepts it identically. The Python sibling (`../python/publish.py`) is a
 * line-for-line mirror.
 *
 * Run: copy `../.env.example` to `../.env`, fill OCTOCHAT_BOT_TOKEN +
 * OCTOCHAT_BOT_SIGN_PATH (and optionally MESSAGE) from the app, then `pnpm start`.
 */
import { join } from 'node:path';

import { generateDeviceKeys, mintDeviceCap } from '@drakkar.software/starfish-identities';
import { signAppendAuthor, signRequest, stableStringify } from '@drakkar.software/starfish-protocol';
import type { SignableMethod } from '@drakkar.software/starfish-protocol';
import { parsePublicLink, redeemPublicLink } from '@drakkar.software/starfish-sharing';

// Load the example's shared `.env` (Node ≥20.12), resolved at the example root
// (two levels up from this file) so it works whether you launch from `ts/` or the
// repo root. Falls back to the real environment when there's no `.env`.
try {
  process.loadEnvFile(join(import.meta.dirname, '..', '..', '.env'));
} catch {
  /* no .env file — rely on exported env vars */
}

interface Config {
  serverUrl: string; // sync base URL (may include a mount path, e.g. `https://host/sync`)
  namespace: string; // bare name, or '' for the root-mounted local dev server
  botToken: string; // stream-bot audience-cap token ("Bot link token")
  signPath: string; // panel "Path to sign", un-namespaced
  message: string; // the text to publish
  name: string; // optional display name to publish as the bot's profile ('' = none)
}

/** Read + validate config. Called from `main`, so a missing var surfaces as a
 *  clean `[publish] fatal: …` line rather than an uncaught module-eval throw. */
function loadConfig(): Config {
  const required = (name: string): string => {
    const v = process.env[name]?.trim();
    if (!v) throw new Error(`Missing required env var ${name} (see .env.example)`);
    return v;
  };
  return {
    serverUrl: process.env.STARFISH_URL?.trim() || 'http://localhost:8787',
    namespace: process.env.STARFISH_NAMESPACE?.trim() || '',
    botToken: required('OCTOCHAT_BOT_TOKEN'),
    signPath: required('OCTOCHAT_BOT_SIGN_PATH'),
    message: process.env.MESSAGE?.trim() || 'Hello from the OctoChat publish example 🐙',
    name: process.env.BOT_NAME?.trim() || '',
  };
}

/** The bot's in-app author id: SHA-256(edPub) first 32 hex, mirroring the SDK's
 *  userId derivation so the stream renders a stable author for the bot's posts. */
async function userIdFromEdPub(edPubHex: string): Promise<string> {
  const bytes = Uint8Array.from(edPubHex.match(/../g)!.map((h) => parseInt(h, 16)));
  const digest = await globalThis.crypto.subtle.digest('SHA-256', bytes);
  return [...new Uint8Array(digest)].map((b) => b.toString(16).padStart(2, '0')).join('').slice(0, 32);
}

/**
 * Publish the bot's PUBLIC PROFILE `{ pseudo }` so the app shows a friendly name
 * instead of the hex author-id prefix. The `profile` collection is public-read
 * but write-gated on `device:root`, granted ONLY to a SELF-SIGNED device cap
 * (`iss === sub`). So the bot mints a device cap over its OWN key (issuer ===
 * subject), which the server admits as `device:root`; the path
 * `user/<authorId>/profile` binds `{identity}` to that same key's user id — the
 * `authorId` it stamps on its message. One signed POST, same wire format as the
 * append below (`{ data, baseHash }`); a fresh per-run key means the doc never
 * exists yet, so `baseHash: null` is a first write.
 */
async function publishProfile(cfg: Config, keys: ReturnType<typeof generateDeviceKeys>, authorId: string): Promise<void> {
  const cap = await mintDeviceCap(
    keys.edPriv,
    keys.edPub,
    { edPubHex: keys.edPub, kemPubHex: keys.kemPub },
    { ops: ['read', 'write', 'list'], collections: ['profile'], paths: [`user/${authorId}/profile`] },
  );
  const actionPath = (cfg.namespace ? `/v1/${cfg.namespace}` : '') + `/push/user/${authorId}/profile`;
  const url = `${cfg.serverUrl.replace(/\/+$/, '')}${actionPath}`;
  const host = new URL(cfg.serverUrl).host;
  const body = JSON.stringify({ data: { pseudo: cfg.name }, baseHash: null });

  const { alg, sig, ts, nonce } = await signRequest(
    { method: 'POST' as SignableMethod, pathAndQuery: actionPath, host, body },
    keys.edPriv,
  );
  const capB64 = Buffer.from(stableStringify(cap as unknown as Record<string, unknown>), 'utf-8').toString('base64');

  const res = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      Authorization: `Cap ${capB64}`,
      'X-Starfish-Sig': sig,
      'X-Starfish-Ts': String(ts),
      'X-Starfish-Nonce': nonce,
      'X-Starfish-Alg': alg,
    },
    body,
  });
  if (!res.ok) {
    throw new Error(`profile write failed: ${res.status} ${await res.text().catch(() => '')}`);
  }
}

async function main(): Promise<void> {
  const cfg = loadConfig();

  // The bot's own keypair redeems the audience-cap token. Fresh per run is fine:
  // the "Connect a bot" panel mints with no `allowedIdentities`, so any key may
  // redeem. To PIN the bot, mint a credential allow-listing this edPub (logged below).
  const keys = generateDeviceKeys();
  const authorId = await userIdFromEdPub(keys.edPub);

  // The path the SERVER observes: namespace prefix + the panel's action path. We
  // sign THIS exact string and POST it under serverUrl (whose mount, e.g. `/sync`,
  // nginx strips before the server sees the signed path) — matching how
  // StarfishClient signs `applyNamespace(path)` while POSTing `baseUrl + that`.
  const actionPath = (cfg.namespace ? `/v1/${cfg.namespace}` : '') + cfg.signPath;
  const url = `${cfg.serverUrl.replace(/\/+$/, '')}${actionPath}`;
  const host = new URL(cfg.serverUrl).host;

  // The OctoChat chat UI reads a typed envelope; a message is `{ t:'msg', e: StoredMsg }`.
  // The server stamps each appended element with an authoritative `{ ts }`, so the
  // `ts` we send is only a client hint (omit it to avoid any monotonic 409).
  const element = {
    t: 'msg',
    e: { id: globalThis.crypto.randomUUID(), authorId, ts: Date.now(), text: cfg.message },
  };

  // Author proof (required by `requireAuthorSignature`, on by default since
  // Starfish alpha.8): sign the element bound to the storage `documentKey` (the
  // un-namespaced `signPath` minus `/push/`) with the redeemer key. `authorPubkey`
  // is the redeemer (= the `X-Starfish-Pub` presenter), so the server's
  // author==presenter check passes. Must precede the request signature below.
  const documentKey = cfg.signPath.replace(/^\/push\//, '');
  const author = signAppendAuthor(documentKey, element, keys.edPub, keys.edPriv);

  // Exact `StarfishClient.append` body: `{ data }` plus the author proof. Bind it
  // ONCE so the signed bytes equal the bytes sent on the wire.
  const body = JSON.stringify({ data: element, ...author });

  const redeemHeaders = await redeemPublicLink(parsePublicLink(cfg.botToken), {
    redeemerEdPrivHex: keys.edPriv,
    redeemerEdPubHex: keys.edPub,
    method: 'POST',
    pathAndQuery: actionPath,
    body, // signed bytes MUST equal the bytes sent on the wire
    host,
  });

  console.log('[publish] OctoChat stream publish');
  console.log(`[publish] server   ${cfg.serverUrl}${cfg.namespace ? `  (namespace ${cfg.namespace})` : '  (local, no namespace)'}`);
  console.log(`[publish] identity ${keys.edPub}  (allow-list this edPub to pin the bot)`);

  // Optional display name: publish the bot's profile pseudo first so the message
  // below renders under a friendly name (not the hex id). Fresh keys per run mean
  // a new profile each run; persist the keypair externally for one identity that
  // survives restarts (the example has no built-in stable-key option).
  if (cfg.name) {
    await publishProfile(cfg, keys, authorId);
    console.log(`[publish] profile  set display name → "${cfg.name}"`);
  }

  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Accept: 'application/json', ...redeemHeaders },
    body,
  });
  if (!res.ok) {
    throw new Error(`append failed: ${res.status} ${await res.text().catch(() => '')}`);
  }
  console.log(`[publish] appended → ${cfg.message}`);
}

main().catch((e) => {
  console.error('[publish] fatal:', (e as Error).message);
  process.exit(1);
});

File: examples/stream-publish-bot/python/publish.py
"""OctoChat stream PUBLISH — a standalone Python script that publishes ONE message
into a PUBLIC stream room as a bot, then exits. No ``/events``, no webhook, no
waiting: just the "post" half of an integration. Line-for-line mirror of the
TypeScript sibling (``../ts/src/publish.ts``).

    publish  ──append──▶  streampub room (access:'public')
              (audience-cap bot token)

A stream room is an append-only log, so posting is a single signed ``POST /push``
— no pull / merge / hash, no sync protocol. The bot is authorized by the
owner-minted "Connect a bot" token, a Starfish ``create_public_link`` AUDIENCE cap
(no embedded secret): the bot generates its OWN keypair and signs the request with
it, naming that key via ``X-Starfish-Pub`` (``redeem_public_link``). A leaked token
is therefore useless to anyone who can't also sign, and writes stay attributable
per bot.

We reproduce ``StarfishClient.append``'s exact wire format
(``body = json({"data": element})``, ``Content-Type: application/json``) so the
server accepts it identically.

The Starfish Python SDK is not yet on PyPI — install it editable from a satellite
source checkout (see ``../README.md``). Then:
    python publish.py
"""

from __future__ import annotations

import base64
import hashlib
import json
import os
import time
import urllib.error
import urllib.request
import uuid
from pathlib import Path
from urllib.parse import urlsplit

from starfish_identities import generate_device_keys, mint_device_cap
from starfish_protocol import sign_request, stable_stringify
from starfish_sharing import parse_public_link, redeem_public_link


def load_env_file(path: Path) -> None:
    """Minimal ``.env`` loader (no python-dotenv dep): ``KEY=value`` lines, ``#``
    comments and blanks skipped. Already-exported env vars win, like most loaders.
    Mirrors the TS side's ``process.loadEnvFile`` so both read the same file."""
    try:
        text = path.read_text(encoding="utf-8")
    except OSError:
        return  # no .env file — rely on exported env vars
    for raw in text.splitlines():
        line = raw.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, _, value = line.partition("=")
        key = key.strip()
        value = value.strip()
        # Strip one layer of matching surrounding quotes, like Node's loadEnvFile.
        if len(value) >= 2 and value[0] == value[-1] and value[0] in ("'", '"'):
            value = value[1:-1]
        if key and key not in os.environ:
            os.environ[key] = value


def required(name: str) -> str:
    """Read a required env var, or exit with a clean ``[publish] fatal: …`` line."""
    value = (os.environ.get(name) or "").strip()
    if not value:
        raise SystemExit(f"[publish] fatal: missing required env var {name} (see .env.example)")
    return value


def user_id_from_ed_pub(ed_pub_hex: str) -> str:
    """The bot's in-app author id: sha256(edPub) first 32 hex, mirroring the SDK's
    userId derivation so the stream renders a stable author for the bot's posts."""
    return hashlib.sha256(bytes.fromhex(ed_pub_hex)).hexdigest()[:32]


def publish_profile(
    server_url: str,
    namespace: str,
    keys: dict[str, str],
    author_id: str,
    pseudo: str,
) -> None:
    """Publish the bot's PUBLIC PROFILE {pseudo} so the app shows a friendly name
    instead of the hex author-id prefix. The ``profile`` collection is public-read
    but write-gated on ``device:root``, granted ONLY to a SELF-SIGNED device cap
    (``iss === sub``). So the bot mints a device cap over its OWN key (issuer ===
    subject), admitted as ``device:root``; the path ``user/<authorId>/profile``
    binds ``{identity}`` to that same key's user id — the ``authorId`` it stamps on
    its message. One signed POST, same wire format as the append below
    (``{data, baseHash}``); a fresh per-run key means the doc never exists yet, so
    ``baseHash=None`` is a first write. Line-for-line mirror of the TS sibling's
    ``publishProfile``."""
    cap = mint_device_cap(
        keys["edPriv"],
        keys["edPub"],
        {"edPubHex": keys["edPub"], "kemPubHex": keys["kemPub"]},
        {"ops": ["read", "write", "list"], "collections": ["profile"], "paths": [f"user/{author_id}/profile"]},
    )
    action_path = (f"/v1/{namespace}" if namespace else "") + f"/push/user/{author_id}/profile"
    url = server_url.rstrip("/") + action_path
    host = urlsplit(server_url).netloc
    body = json.dumps({"data": {"pseudo": pseudo}, "baseHash": None}, separators=(",", ":")).encode("utf-8")

    signature = sign_request("POST", action_path, body, keys["edPriv"], host=host)
    cap_b64 = base64.b64encode(stable_stringify(cap).encode("utf-8")).decode("ascii")
    headers = {
        "Content-Type": "application/json",
        "Accept": "application/json",
        "Authorization": f"Cap {cap_b64}",
        "X-Starfish-Sig": signature.sig,
        "X-Starfish-Ts": str(signature.ts),
        "X-Starfish-Nonce": signature.nonce,
        "X-Starfish-Alg": signature.alg,
    }
    request = urllib.request.Request(url, data=body, headers=headers, method="POST")
    try:
        with urllib.request.urlopen(request) as response:
            response.read()
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", "replace")
        raise SystemExit(f"[publish] fatal: profile write failed: {exc.code} {detail}")
    except urllib.error.URLError as exc:
        raise SystemExit(f"[publish] fatal: {exc.reason}")


def main() -> None:
    # Load the example's shared `.env` at the example root (one level up from here).
    load_env_file(Path(__file__).resolve().parent.parent / ".env")

    server_url = (os.environ.get("STARFISH_URL") or "").strip() or "http://localhost:8787"
    namespace = (os.environ.get("STARFISH_NAMESPACE") or "").strip()
    bot_token = required("OCTOCHAT_BOT_TOKEN")
    sign_path = required("OCTOCHAT_BOT_SIGN_PATH")
    message = (os.environ.get("MESSAGE") or "").strip() or "Hello from the OctoChat publish example 🐙"
    name = (os.environ.get("BOT_NAME") or "").strip()

    # The bot's own keypair redeems the audience-cap token. Fresh per run is fine:
    # the "Connect a bot" panel mints with no allowed identities, so any key may
    # redeem. To PIN the bot, mint a credential allow-listing this edPub (logged below).
    keys = generate_device_keys()
    author_id = user_id_from_ed_pub(keys["edPub"])

    # The path the SERVER observes: namespace prefix + the panel's action path. We
    # sign THIS exact string and POST it under server_url (whose mount, e.g. `/sync`,
    # nginx strips before the server sees the signed path) — matching how
    # StarfishClient signs `applyNamespace(path)` while POSTing `baseUrl + that`.
    action_path = (f"/v1/{namespace}" if namespace else "") + sign_path
    url = server_url.rstrip("/") + action_path
    host = urlsplit(server_url).netloc

    # The OctoChat chat UI reads a typed envelope; a message is { t:'msg', e: StoredMsg }.
    # The server stamps each appended element with an authoritative { ts }; the `ts`
    # we send is only a client hint.
    element = {
        "t": "msg",
        "e": {"id": str(uuid.uuid4()), "authorId": author_id, "ts": int(time.time() * 1000), "text": message},
    }

    # Exact StarfishClient.append body: the element wrapped as { data }. Bind it ONCE
    # so the signed bytes equal the bytes sent on the wire.
    body = json.dumps({"data": element}, separators=(",", ":")).encode("utf-8")

    headers = redeem_public_link(
        parse_public_link(bot_token),
        redeemer_ed_priv_hex=keys["edPriv"],
        redeemer_ed_pub_hex=keys["edPub"],
        method="POST",
        path_and_query=action_path,
        body=body,  # signed bytes MUST equal the bytes sent on the wire
        host=host,
    )
    headers["Content-Type"] = "application/json"
    headers["Accept"] = "application/json"

    print("[publish] OctoChat stream publish")
    print(f"[publish] server   {server_url}" + (f"  (namespace {namespace})" if namespace else "  (local, no namespace)"))
    print(f"[publish] identity {keys['edPub']}  (allow-list this edPub to pin the bot)")

    # Optional display name: publish the bot's profile pseudo first so the message
    # below renders under a friendly name (not the hex id). Fresh keys per run mean
    # a new profile each run; persist the keypair externally for one identity that
    # survives restarts (the example has no built-in stable-key option).
    if name:
        publish_profile(server_url, namespace, keys, author_id, name)
        print(f'[publish] profile  set display name → "{name}"')

    request = urllib.request.Request(url, data=body, headers=headers, method="POST")
    try:
        with urllib.request.urlopen(request) as response:
            response.read()
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", "replace")
        raise SystemExit(f"[publish] fatal: append failed: {exc.code} {detail}")
    except urllib.error.URLError as exc:
        raise SystemExit(f"[publish] fatal: {exc.reason}")

    print(f"[publish] appended → {message}")


if __name__ == "__main__":
    main()
